tortoise/tortoise.go

```diff
@@ -415,33 +418,33 @@ func (t *turtle) countBallot(ballot *ballotInfo) error {
 
 func (t *turtle) verifyLayers() {
 	var (
-		verified           = maxLayer(t.evicted, types.GetEffectiveGenesis())
+		verified           = max(t.evicted, types.GetEffectiveGenesis())
 		nverified, changed types.LayerID
 	)
 	if t.isFull {
 		nverified, changed = t.runFull()
 		vverified, vchanged := t.runVerifying()
 		if nverified == t.processed-1 && nverified == vverified {
 			t.switchModes()
-			changed = types.MinLayer(changed, vchanged)
+			changed = min(changed, vchanged)
```

Comment:
Previously if `changed` was 0, it would pick `vchanged`. Now it will pick `changed`. Is it intended?

Reply:
The problem with the previous implementation was that 0 had to meanings: 0 (as a minimum) and no minimum.

`t.runFull()` and `t.runVerifying` would return 0 for `changed` and `vchanged` if no changes were detected. Now it instead returns `math.MaxUint32`.

This removes the need for handling 0 in a special way (within `types.MinLayer`) and instead the detection for "no changed layer" is now asserting that `changed` is not `math.MaxUint32`.

Reply:
On top of that the documentation for `LayerID` fits the suggested implementation rather than the current one:

```
// LayerID is representing a layer number. Zero value is safe to use, and means 0.
// Internally it is a simple wrapper over uint32 and should be considered immutable
// the same way as any integer.
```

But the code assumes that 0 means "no layer" rather than the first layer. Also `types.MinLayer` is not transitive:

```go
types.MinLayer(a, b) <= types.MinLayer(b, c) // this can be false if b == 0
min(a, b) <= min(b, c) // this is always true
```

Reply:
I know, I'm just worried that it could break some code that depends on this behaviour

Reply:
> t.runFull() and t.runVerifying would return 0 for changed and vchanged if no changes were detected. Now it instead returns math.MaxUint32.

How about changing these functions to return a pointer, where `nil` means "not found" instead? It has the benefit of making it clear that the function returns an optional value. The downside is that it would most likely complicate the code using them :(

Reply:
I think the implementation with `math.MaxUint32` is already simpler than the current implementation?

> I know, I'm just worried that it could break some code that depends on this behaviour

The methods `runFull` and `runVerifying` only have package visibility and are only called in `verifyLayers`. Their returned values are only used in that method. For the only caller of `verifyLayers` (`onLayer`) nothing changed.

All 4 methods are well tested (100% coverage) - the other methods in that file also have near 100% coverage; unit and system tests pass. Earlier I accidentally had `max` where `min` was supposed to go and multiple tests failed.

I appreciate your concern, but think the change is solid 🙂